packages/js/src/analysis/plugins/shortcode-plugin.js

```diff
@@ -0,0 +1,324 @@
+/* global tinyMCE */
+/* global wpseoScriptData */
+/* global ajaxurl */
+/* global _ */
+
+const shortcodeNameMatcher = "[^<>&/\\[\\]\x00-\x20=]+?";
+const shortcodeAttributesMatcher = "( [^\\]]+?)?";
+
+const shortcodeStartRegex = new RegExp( "\\[" + shortcodeNameMatcher + shortcodeAttributesMatcher + "\\]", "g" );
+const shortcodeEndRegex = new RegExp( "\\[/" + shortcodeNameMatcher + "\\]", "g" );
+
+/**
+ * The Yoast Shortcode plugin parses the shortcodes in a given piece of text. It analyzes multiple input fields for
+ * shortcodes which it will preload using AJAX.
+ */
+class YoastShortcodePlugin {
+	/**
+	 * Constructs the YoastShortcodePlugin.
+	 *
+	 * @property {RegExp} keywordRegex Used to match a given string for valid shortcode keywords.
+	 * @property {RegExp} closingTagRegex Used to match a given string for shortcode closing tags.
+	 * @property {RegExp} nonCaptureRegex Used to match a given string for non-capturing shortcodes.
+	 * @property {Array} parsedShortcodes Used to store parsed shortcodes.
+	 *
+	 * @param {Object}   interface                      Object Formerly Known as App, but for backwards compatibility
+	 *                                                  still passed here as one argument.
+	 * @param {function} interface.registerPlugin       Register a plugin with Yoast SEO.
+	 * @param {function} interface.registerModification Register a modification with Yoast SEO.
+	 * @param {function} interface.pluginReady          Notify Yoast SEO that the plugin is ready.
+	 * @param {function} interface.pluginReloaded       Notify Yoast SEO that the plugin has been reloaded.
+	 * @param {Array}	shortcodesToBeParsed The array of shortcodes to be parsed.
+	 * @returns {void}
+	 */
+	constructor( { registerPlugin, registerModification, pluginReady, pluginReloaded }, shortcodesToBeParsed ) {
+		this._registerModification = registerModification;
+		this._pluginReady = pluginReady;
+		this._pluginReloaded = pluginReloaded;
+
+		registerPlugin( "YoastShortcodePlugin", { status: "loading" } );
+		this.bindElementEvents();
+
+		const keywordRegexString = "(" + shortcodesToBeParsed.join( "|" ) + ")";
+
+		// The regex for matching shortcodes based on the available shortcode keywords.
+		this.keywordRegex = new RegExp( keywordRegexString, "g" );
```

Comment:
Let's rename these variables, as `keyword` normally has another meaning in our application.